src/IO/CachedInMemoryReadBufferFromFile.cpp

```diff
@@ -41,9 +42,28 @@ String CachedInMemoryReadBufferFromFile::getInfoForLog()
 
 bool CachedInMemoryReadBufferFromFile::isSeekCheap()
 {
-    /// If working buffer is empty then the next nextImpl() call will have to send a new read
-    /// request anyway, seeking doesn't make it more expensive.
-    return available() == 0 || last_read_hit_cache;
+    /// Seek is cheap in the sense that seek()+nextImpl() is never much slower than ignore()+nextImpl()
+    /// (which is what the caller cares about).
+    return true;
+}
+
+bool CachedInMemoryReadBufferFromFile::isContentCached(size_t offset, size_t /*size*/)
+{
+    /// Usually this is called immediately after seek()ing to `offset`.
+
+    if (!working_buffer.empty())
+    {
+        chassert(chunk);
+        return chunk->key.offset <= offset && chunk->key.offset + chunk->key.size > offset;
+    }
+
+    size_t block_size = settings.page_cache_block_size;
+    cache_key.offset = offset / block_size * block_size;
+    cache_key.size = std::min(block_size, file_size.value() - cache_key.offset);
+
+    chunk = cache->get(cache_key, settings.page_cache_inject_eviction);
```

Comment:
why not contains?

Reply:
To avoid doing the lookup twice (here and in subsequent nextImpl).